### PR TITLE
Follow hashes: new syntax

### PR DESCRIPTION
**Pull request by rafie**

### CODE REVIEW

File: src/rules.c [hunk 1]
```diff
@@ -71,6 +66,52 @@ void SchemaRules_RemoveSpecRules(IndexSpec *spec) {
...
+
+RSLanguage SchemaRule_HashLang(const SchemaRule *rule, RedisModuleKey *key, const char *kname) {
+  RedisModuleString *lang_rms = NULL;
+  if (!rule->lang_field) {
+    return DEFAULT_LANGUAGE;
+  }
+  int rv = RedisModule_HashGet(key, REDISMODULE_HASH_CFIELDS, rule->lang_field, &lang_rms, NULL);
+  if (rv != REDISMODULE_OK) {
+    RedisModule_Log(NULL, "warning", "invalid field %s for key %s", rule->lang_field, kname);
+    return DEFAULT_LANGUAGE;
```
Comment (ashtul) [inline src/rules.c:79]: Why not return `RS_LANG_UNSUPPORTED`?

File: src/rules.c [hunk 2]
```diff
@@ -71,6 +66,52 @@ void SchemaRules_RemoveSpecRules(IndexSpec *spec) {
...
+  }
+  const char *lang_s = (const char *) RedisModule_StringPtrLen(lang_rms, NULL); 
+  RSLanguage lang = RSLanguage_Find(lang_s);
+  if (lang == RS_LANG_UNSUPPORTED) {
+    RedisModule_Log(NULL, "warning", "invalid language for for key %s", kname);
+    return DEFAULT_LANGUAGE;
```
Comment (ashtul) [inline src/rules.c:85]: Why not return `RS_LANG_UNSUPPORTED`?

File: src/aggregate/functions/rules.c [hunk 3]
```diff
@@ -2,19 +2,3 @@
 #include "aggregate/expr/expression.h"
```
Comment (ashtul) [inline src/aggregate/functions/rules.c:2]: Remove file

File: src/dep/triemap/triemap.c [hunk 4]
```diff
@@ -266,6 +266,56 @@ void *TrieMapNode_Find(TrieMapNode *n, char *str, tm_len_t len) {
...
+    tm_len_t node_offset = 0;
+    tm_len_t nlen = node->len;
+    while (offset < len && node_offset < nlen && str[offset] == node->str[node_offset]) {
+      offset++;
+      node_offset++;
+    }
+
+    // no match
+    if (node_offset != nlen) {
+      return array_len(*results);
```
Comment (ashtul) [inline src/dep/triemap/triemap.c:282]: return 0

File: src/dep/triemap/triemap.c [hunk 5]
```diff
@@ -266,6 +266,56 @@ void *TrieMapNode_Find(TrieMapNode *n, char *str, tm_len_t len) {
...
+    }
+
+    // at the end of both strings
+    if (offset == len) {
+      // If this is a terminal, non deleted node
+      if (__trieMapNode_isTerminal(node) && !__trieMapNode_isDeleted(node)) {
+        *results = array_ensure_append_1(*results, node->value);
+      }
+      return array_len(*results);
```
Comment (ashtul) [inline src/dep/triemap/triemap.c:291]: return 1

File: src/rules.c [hunk 6]
```diff
@@ -1,46 +1,173 @@
...
-  // TODO: Remove from global list
+  TrieMapIterator_Free(it); 
 }
+
+//---------------------------------------------------------------------------------------------
+
+SchemaPrefixNode *SchemaPrefixNode_Create(const char *prefix, IndexSpec *index) {
+  SchemaPrefixNode *node = rm_calloc(1, sizeof(*node));
+  node->prefix = prefix;
+  node->index_specs = NULL;
```
Comment (ashtul) [inline src/rules.c:164]: remove

File: src/rules.c [hunk 7]
```diff
@@ -1,46 +1,173 @@
...
+    char *p;
+    tm_len_t len;
+    SchemaPrefixNode *node;
+    if (!TrieMapIterator_Next(it, &p, &len, (void **) &node)) {
+      break;
+    }
+    int j = -1;
+    for (int i = 0; i < array_len(node->index_specs); ++i) {
+      if (node->index_specs[i] == spec) {
+        j = i;
```
Comment (ashtul) [inline src/rules.c:148]: why not call `array_del_fast` here?

File: src/spec.c [hunk 8]
```diff
@@ -492,11 +498,15 @@ IndexSpec *IndexSpec_Parse(const char *name, const char **argv, int argc, QueryE
...
+  if (rule_args.filter_exp_str || rule_prefixes.argc > 0) {
+    rule_args.nprefixes = rule_prefixes.argc;
+    rule_args.prefixes = (const char **) rule_prefixes.objs;
+    spec->rule = SchemaRule_Create(&rule_args, spec, status);
+    if (!spec->rule) {
       goto failure;
     }
-    SchemaRules_g = array_ensure_append(SchemaRules_g, &spec, 1, IndexSpec *);
+  } else {
+    spec->rule = NULL;
```
Comment (ashtul) [inline src/spec.c:509]: `goto failure`?

File: src/spec.c [hunk 9]
```diff
@@ -1372,22 +1383,51 @@ int IndexSpec_RegisterType(RedisModuleCtx *ctx) {
...
+  }
+
+#ifdef DEBUG
+  const char *key_p = RedisModule_StringPtrLen(key, &n);
+#endif
   RedisSearchCtx sctx = SEARCH_CTX_STATIC(ctx, spec);
   Document doc = {0};
-  // TODO: get language and score from spec->rule->setting
-  Document_Init(&doc, key, 1, DEFAULT_LANGUAGE);
+  Document_Init(&doc, key, 0, DEFAULT_LANGUAGE);
```
Comment (ashtul) [inline src/spec.c:1396]: `1` was good

File: src/rules.c [hunk 10]
```diff
@@ -1,46 +1,185 @@
...
+SchemaRule *SchemaRule_Create(SchemaRuleArgs *args, IndexSpec *spec, QueryError *status) {
+  SchemaRule *rule = rm_calloc(1, sizeof(*rule));
 
-    if (rulesopts->score) {
-      rule->setting.score = rm_strdup(rulesopts->score);
-    }
-    
-    if (rulesopts->lang) {
-      rule->setting.lang = rm_strdup(rulesopts->lang);
+  rule->type = rm_strdup(args->type);
```
Comment (ashtul) [inline src/rules.c:13]: Use enum

Comment (rafie) [reply src/rules.c:13]: @ashtul it's not just the enum, when we support more follow types, we'll have to differentiate on how we perform it. Right now, the type is just a word - let's keep it like this for simplicity and do the extra abstraction when we're more stable.

**Pull request merged**